Panic in isReturnedWithNonNilError

@ryanhall07 ran into this and debugged it to `isReturnedWithNonNilError`, I was able to come up with a small repro:

```go
type Params struct {
        Name string // required
}

func foo() {
        f := func() string {
                return Params{}.Name
        }
        fmt.Println(f())
}
```

Panic:
```
$ requiredfield ./...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x100271e60]

goroutine 600 [running]:
go.abhg.dev/requiredfield.(*enforcer).isReturnedWithNonNilError(0x10044c2e0?, {0x14007068400?, 0x14007af83c0?, 0x4?})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:167 +0x140
go.abhg.dev/requiredfield.(*enforcer).visit(0x14003dacc30, {0x1004a5c40?, 0x14007b285c0}, {0x14007068400, 0x9, 0x10})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:87 +0x228
go.abhg.dev/requiredfield.run.(*enforcer).Enforce.func1({0x1004a5c40?, 0x14007b285c0?}, 0x0?, {0x14007068400?, 0x10044b6e0?, 0x0?})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:31 +0x3c
# truncated, see details for full panic
```

<details>

```
$ requiredfield ./...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x100271e60]

goroutine 600 [running]:
go.abhg.dev/requiredfield.(*enforcer).isReturnedWithNonNilError(0x10044c2e0?, {0x14007068400?, 0x14007af83c0?, 0x4?})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:167 +0x140
go.abhg.dev/requiredfield.(*enforcer).visit(0x14003dacc30, {0x1004a5c40?, 0x14007b285c0}, {0x14007068400, 0x9, 0x10})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:87 +0x228
go.abhg.dev/requiredfield.run.(*enforcer).Enforce.func1({0x1004a5c40?, 0x14007b285c0?}, 0x0?, {0x14007068400?, 0x10044b6e0?, 0x0?})
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:31 +0x3c
golang.org/x/tools/go/ast/inspector.(*Inspector).WithStack(0x14007b24588, {0x1006799d0?, 0x303dacc28?, 0x147873dc0?}, 0x14003dacc48)
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/ast/inspector/inspector.go:179 +0x160
go.abhg.dev/requiredfield.(*enforcer).Enforce(...)
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/enforce.go:26
go.abhg.dev/requiredfield.run(0x14007196b60)
        /Users/prashant/go/pkg/mod/go.abhg.dev/requiredfield@v0.5.0/analyzer.go:64 +0x13c
golang.org/x/tools/go/analysis/checker.(*Action).execOnce.func3(...)
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/analysis/checker/checker.go:367
golang.org/x/tools/go/analysis/checker.(*Action).execOnce(0x140060f78c0)
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/analysis/checker/checker.go:388 +0x920
sync.(*Once).doSlow(0x0?, 0x10000010067f460?)
        /Users/prashant/.local/share/mise/installs/go/1.24.5/src/sync/once.go:78 +0xf0
sync.(*Once).Do(...)
        /Users/prashant/.local/share/mise/installs/go/1.24.5/src/sync/once.go:69
golang.org/x/tools/go/analysis/checker.(*Action).exec(...)
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/analysis/checker/checker.go:265
golang.org/x/tools/go/analysis/checker.execAll.func1(0x1004a65a0?)
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/analysis/checker/checker.go:253 +0x50
created by golang.org/x/tools/go/analysis/checker.execAll in goroutine 1
        /Users/prashant/go/pkg/mod/golang.org/x/tools@v0.30.0/go/analysis/checker/checker.go:259 +0x14c
```

</details>

It seems like there's a couple of issues:
 * Functions that don't have returns have `ftype.Results = nil`, which panics here:
https://github.com/abhinav/requiredfield/blob/8f49d1f3bebbe797a2c8c33dc548d1192edd5ad5/enforce.go#L166-L168

 * It seems like the return --> matching function signature matching is incorrect since the return is inside of `func() string` but it seems to be looking at `func foo()`

--- a/.changes/unreleased/Fixed-20250811-163138.yaml
+++ b/.changes/unreleased/Fixed-20250811-163138.yaml
@@ -1,0 +1,3 @@
+kind: Fixed
+body: Fix panic when returning a struct from a nested function literal.
+time: 2025-08-11T16:31:38.812525-07:00

--- a/enforce.go
+++ b/enforce.go
@@ -156,11 +156,11 @@ func (e *enforcer) isReturnedWithNonNilError(stack []ast.Node) bool {
 	}
 
 	// The last return type must be an error.
-	returnTypes := ftype.Results.List
-	if len(returnTypes) == 0 {
+	if ftype.Results == nil || len(ftype.Results.List) == 0 {
 		// No return types.
 		return false
 	}
+	returnTypes := ftype.Results.List
 	lastReturnType, ok := returnTypes[len(returnTypes)-1].Type.(*ast.Ident)
 	if !ok || lastReturnType.Name != "error" {
 		// The last return type is not "error".

--- a/enforce.go
+++ b/enforce.go
@@ -149,16 +149,8 @@ func (e *enforcer) isReturnedWithNonNilError(stack []ast.Node) bool {
 	}
 
 	// Find the nearest function's type for the return statement.
-	var ftype *ast.FuncType
-	for idx := retIdx - 1; idx >= 0; idx-- {
-		switch n := stack[idx].(type) {
-		case *ast.FuncDecl:
-			ftype = n.Type
-		case *ast.FuncLit:
-			ftype = n.Type
-		}
-	}
-	if ftype == nil {
+	ftype, ok := parentFuncType(stack[:retIdx])
+	if !ok || ftype == nil {
 		// Impossible, but we don't want to panic.
 		return false
 	}
@@ -199,4 +191,17 @@ func (e *enforcer) isReturnedWithNonNilError(stack []ast.Node) bool {
 
 	// Target is not part of the last return value.
 	return true
+}
+
+func parentFuncType(stack []ast.Node) (*ast.FuncType, bool) {
+	for idx := len(stack) - 1; idx >= 0; idx-- {
+		switch n := stack[idx].(type) {
+		case *ast.FuncDecl:
+			return n.Type, true
+		case *ast.FuncLit:
+			return n.Type, true
+		}
+	}
+
+	return nil, false
 }

--- a/testdata/src/e/e.go
+++ b/testdata/src/e/e.go
@@ -58,6 +58,12 @@ var functionLiteral = func() (Foo, error) {
 	}
 }
 
+func nestedFunctionLiteral() {
+	_ = func() Foo {
+		return Foo{} // want "missing required fields: Bar"
+	}
+}
+
 func errorIsNotLastReturn() (error, Foo) {
 	if rand.Int()%2 == 0 {
 		return errors.New("fail"), Foo{} // want "missing required fields: Bar"

